pkg/controller/volume/expand/cache/volume_resize_map.go

```diff
@@ -0,0 +1,303 @@
+}
+
+func (resizeMap *volumeResizeMap) MarkForFileSystemResize(pvcr *PvcWithResizeRequest, newSize resource.Quantity) error {
+	resizeMap.Lock()
+	defer resizeMap.Unlock()
+
+	err := resizeMap.updatePvSize(pvcr, newSize)
+	if err != nil {
+		glog.V(4).Infof("Error updating PV spec capacity for volume %q with : %v", pvcr.QualifiedName(), err)
+		resizeMap.addBackResizeRequest(pvcr)
+		return err
+	}
+	return nil
+}
+
+func (resizeMap *volumeResizeMap) updatePvSize(pvcr *PvcWithResizeRequest, newSize resource.Quantity) error {
+	oldPv := pvcr.VolumeSpec.PersistentVolume
+	clone, err := scheme.Scheme.DeepCopy(oldPv)
+
+	if err != nil {
+		return fmt.Errorf("Error cloning PV %q with error : %v", oldPv.Name, err)
+	}
+	pvClone, ok := clone.(*v1.PersistentVolume)
+
+	if !ok {
+		return fmt.Errorf("Unexpected cast error for PV : %v", pvClone)
+	}
+
+	pvClone.Spec.Capacity[v1.ResourceStorage] = newSize
+	_, updateErr := resizeMap.kubeClient.CoreV1().PersistentVolumes().Update(pvClone)
+
+	if updateErr != nil {
+		glog.V(4).Infof("Error updating pv %q with error : %v", pvClone.Name, updateErr)
+		return updateErr
+	}
+	return nil
+}
+
+func (resizeMap *volumeResizeMap) updatePvcCapacity(pvcr *PvcWithResizeRequest, newSize resource.Quantity) error {
+	claimClone, err := util.ClonePVC(pvcr.PVC)
+	if err != nil {
+		return err
+	}
+
+	claimClone.Status.Capacity[v1.ResourceStorage] = newSize
+	_, updateErr := resizeMap.kubeClient.Core().PersistentVolumeClaims(claimClone.Namespace).UpdateStatus(claimClone)
+	if updateErr != nil {
+		glog.V(4).Infof("updating PersistentVolumeClaim[%s] status: failed: %v", pvcr.QualifiedName(), updateErr)
+		return updateErr
+	}
+	return nil
+}
+
+func (resizeMap *volumeResizeMap) updatePvcCapacityAndConditions(pvcr *PvcWithResizeRequest, newSize resource.Quantity, pvcConditions []v1.PersistentVolumeClaimCondition) error {
+
+	claimClone, err := util.ClonePVC(pvcr.PVC)
+	if err != nil {
+		return err
+	}
+
+	claimClone.Status.Capacity[v1.ResourceStorage] = newSize
+	claimClone.Status.Conditions = pvcConditions
+	_, updateErr := resizeMap.kubeClient.CoreV1().PersistentVolumeClaims(claimClone.Namespace).UpdateStatus(claimClone)
+	if updateErr != nil {
+		glog.V(4).Infof("updating PersistentVolumeClaim[%s] status: failed: %v", pvcr.QualifiedName(), updateErr)
+		return updateErr
+	}
+	return nil
+}
+
+// add back the resize request because it may have failed. Assumes that caller has necessary locks.
+func (resizeMap *volumeResizeMap) addBackResizeRequest(pvcr *PvcWithResizeRequest) {
+	pvcUniqueName := pvcr.UniquePvcKey()
+	if _, ok := resizeMap.pvcrs[pvcUniqueName]; ok {
+		glog.V(5).Infof("Found another resize request pending for volume %s", pvcr.QualifiedName())
+		return
+	}
+	pvc, err := resizeMap.kubeClient.CoreV1().PersistentVolumeClaims(pvcr.PVC.Namespace).Get(pvcr.PVC.Name, metav1.GetOptions{})
+	if err != nil {
+		glog.Errorf("Error fetching pvc %v/%v for resize with error : %v", pvc.Namespace, pvc.Name, err)
```

Comment:
This will happen when someone deletes the PVC during resize - say `plugin.ExpandVolumeDevice` fails and the request should be enqueued again and it ends up here. A lower log level would be nice + some comment that  this is expected behavior. It's not obvious that seemingly endless loop of `ExpandVolumeDevice` error + `addBackResizeRequest` for deleted volumes stops here.

Reply:
ack done.